crowbar_framework/app/models/api/node.rb

```diff
@@ -16,6 +16,98 @@
 
 module Api
   class Node < Tableless
+    def initialize(name = nil)
+      @node = NodeObject.find_node_by_name name
+    end
+
+    def upgraded?
+      # FIXME: check this by looking at some upgraded-ok file on the node
+      current_platform = "#{@node[:platform]}-#{@node[:platform_version]}"
+      @upgraded ||= current_platform == @node[:target_platform]
+    end
+
+    def pre_upgrade
+      # TODO: save the global status info about this substep (we started prepare for upgrade)
+
+      # Migrate out the l3 agents and shut down pacemaker
+      script = "/usr/sbin/crowbar-pre-upgrade.sh"
+      out = @node.run_ssh_cmd(script)
```

Comment:
I think it would be worth running it inside `screen(1)`, so that if the ssh pipe fails unexpectedly, at least the script should still stand a good chance of finishing cleanly, which would make it more likely that resuming the upgrade later will succeed. I presume that wrapping inside `screen` should still preserve the exit code status.